base/io.jl

```
@@ -864,11 +864,10 @@ end
 
 function write(io::IO, c::Char)
     u = bswap(reinterpret(UInt32, c))
-    n = 1
+    n = 0
     while true
-        write(io, u % UInt8)
+        n += write(io, u % UInt8)
         (u >>= 8) == 0 && return n
```

This currently unconditionally advances the written character, but what happens in case the first `write` fails, and the second succeeds? Now there's suddenly a torn write involved here, and even though you can theoretically know that not all of the given `Char` has been written (e.g. getting a return value of `3` when a 4-byte `Char` is passed), you still wouldn't know which byte was missed.

---

Is there any Julia IO type where writing a byte can fail, return zero, and then succeed, without some error being thrown?

---

For example, writing a byte with TranscodingStreams.jl will either [return 1 or throw an error](https://github.com/JuliaIO/TranscodingStreams.jl/blob/76c44c59f9015842a5c23bcd90082421b710c26f/src/stream.jl#L480C1-L486C4).

---

Sure, a non-blocking buffered IO whose buffer is temporarily full, for example. I don't know whether there currently is such a type in the ecosystem, but the point is that it could exist and would be a valid `IO`, as far as I can tell.

---

Here's a (slightly contrived) example:

```julia-repl
julia> io = IOBuffer(; maxsize=1)
IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=1, ptr=1, mark=-1)

julia> write(io, 'a')
1

julia> write(io, 'a') # should be 0 with this PR, since the write doesn't succeed
1

julia> seekstart(io); # simulate a read-end on some other process, for example

julia> read(io, Char) # happens on the read-end
'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)

julia> write(io, 'a') # continue writing
1
```

It's a bit awkward to do this with an `IOBuffer`, but the principle is the same for some `IO` type that has an actual read-end that's distinct from the write end. For arbitrary I/O, it's usually preferrable to drop data on the write end and retry later once the buffer is ready to send again. With the current behavior, the writer wouldn't know what to try to retransmit over the I/O, since it's impossible to know which byte(s) of the `Char` was/were not transmitted correctly. Effectively, the number returned by `write` becomes irrelevant, and only matters when it matches `sizeof(Char)` - at which point we might as well only return `true`/`false`. If we instead abort as soon as any internal `write` fails, we know that at least a correct prefix of the `Char` (or any data, in the general case) was returned, and we can retry with _only_ the data that we haven't attempted to transmit at all yet.

---

It really does seem that `write(::IO, ::UInt8)` should throw an error any time it can't write the byte; zero is not a "valid" return value in that case. You can only get zero if you write e.g. an empty string.

An error might sound bad if a device is only temporarily unavailable, but returning zero isn't much better. You can't retry later without some kind of synchronization anyway.

---

I don't think write ever errors for us, given asyncio and other stuff?

---

It does, e.g. trying to write to a read-only stream.

---

> zero is not a "valid" return value in that case.

There is interesting historical data suggesting that some implementations of libc `write` were indeed able to return `0`: https://stackoverflow.com/a/41970485

For quite a bunch of kinds of files, the behavior is unspecified, so more or less anything goes either way :shrug: 

> I don't think write ever errors for us, given asyncio and other stuff?

Right, and for a non-blocking buffered IO it would be incredibly awkward to throw actual errors just because it's full. That possibility would be incredibly detrimental in the common case of success. I admit having `0` signal that is quite a bad API though. I guess this is yet-another case something like a `Result{Int, Err}` sum type would be nice, to distinguish success from errors :thinking:

---

This conversation is worth continuing, but for the purposes of fixing this bug I think it's orthogonal.

Our `AbstractArray` write method can also suffer "torn" writes in the same way:
```julia
function unsafe_write(s::IO, p::Ptr{UInt8}, n::UInt)
    written::Int = 0
    for i = 1:n
        written += write(s, unsafe_load(p, i))
    end
    return written
end
```

This is probably worth splitting into a separate issue and fixing across-the-board. The only thing I think this needs to merge @gbaraldi is a test.

---

Filed https://github.com/JuliaLang/julia/issues/57011 to continue discussion